File: src/main/kotlin/samlang/util/UnionFind.kt
package samlang.util

/**
 * An lazily allocated union find data structure.
 */
class UnionFind {

    private val parent: MutableMap<Int, Int> = hashMapOf()
    private val treeSize: MutableMap<Int, Int> = hashMapOf()

    override fun toString(): String = "[parent: $parent, treeSize: $treeSize]"

    private fun getParent(i: Int): Int {
        val currentParent = parent[i]
        if (currentParent == null) {
            parent[i] = i
            treeSize[i] = 1
            return i
        }
        return currentParent
    }

    private fun getTreeSize(i: Int): Int {
        val currentTreeSize = treeSize[i]
        if (currentTreeSize == null) {
            parent[i] = i
            treeSize[i] = 1
            return 1
        }
        return currentTreeSize
    }

    /**
     * Find the root of [i].
     */
    fun find(i: Int): Int {
        val currentParent = getParent(i)
        return if (currentParent == i) {
            currentParent
        } else {
            val parentOfParent = find(i = currentParent)
            parent[i] = parentOfParent
            parentOfParent
        }
    }

    /**
     * Link [i] and [j] and returns their new common root.
     */
    fun link(i: Int, j: Int): Int {
        var iRoot = find(i = i)
        var jRoot = find(i = j)
        if (iRoot == jRoot) {
            return iRoot
        }
        if (getTreeSize(iRoot) < getTreeSize(jRoot)) {
            val temp = iRoot
            iRoot = jRoot
            jRoot = temp
        }
        parent[jRoot] = iRoot
        treeSize[iRoot] = getTreeSize(iRoot) + getTreeSize(jRoot)
        return iRoot
    }

}

File: samlang-core-utils/union-find.ts
/** An lazily allocated union find data structure. */
export default class UnionFind {
  private readonly parent: Map<number, number> = new Map();

  private readonly treeSize: Map<number, number> = new Map();

  getParent(index: number): number {
    const currentParent = this.parent.get(index);
    if (currentParent == null) {
      this.parent.set(index, index);
      this.treeSize.set(index, 1);
      return index;
    }
    return currentParent;
  }

  getTreeSize(index: number): number {
    const currentTreeSize = this.treeSize.get(index);
    if (currentTreeSize == null) {
      this.parent.set(index, index);
      this.treeSize.set(index, 1);
      return 1;
    }
    return currentTreeSize;
  }

  isLinked(i: number, j: number): boolean {
    return this.findRoot(i) === this.findRoot(j);
  }

  findRoot(index: number): number {
    const currentParent = this.getParent(index);
    if (currentParent === index) {
      return currentParent;
    }
    const parentOfParent = this.findRoot(currentParent);
    this.parent.set(index, parentOfParent);
    return parentOfParent;
  }

  link(i: number, j: number): number {
    let iRoot = this.findRoot(i);
    let jRoot = this.findRoot(j);
    if (iRoot === jRoot) {
      return iRoot;
    }
    if (this.getTreeSize(iRoot) < this.getTreeSize(jRoot)) {
      const temp = iRoot;
      iRoot = jRoot;
      jRoot = temp;
    }
    this.parent.set(jRoot, iRoot);
    this.treeSize.set(iRoot, this.getTreeSize(iRoot) + this.getTreeSize(jRoot));
    return iRoot;
  }
}
